src/hooks/transactions/index.tsx

```diff
@@ -79,15 +84,35 @@ export function TransactionsUpdater() {
   )
   const onReceipt = useCallback(
     ({ chainId, hash, receipt }) => {
+      if (receipt?.status === 0) {
+        onTxFail?.(new Error('Transaction failed'), receipt)
+      } else {
+        onTxSuccess?.(receipt.transactionHash, receipt)
+      }
       updateTxs((txs) => {
         const tx = txs[chainId]?.[hash]
         if (tx) {
           tx.receipt = receipt
         }
       })
     },
-    [updateTxs]
+    [updateTxs, onTxFail, onTxSuccess]
   )
 
+  const oldTxs = useRef({})
+  useEffect(() => {
+    const newTxHashes = Object.keys(pendingTransactions)
+    const oldTxHashes = Object.keys(oldTxs.current)
+    if (newTxHashes.length !== oldTxHashes.length) {
+      // if added new tx
+      newTxHashes.forEach((txHash) => {
+        if (!oldTxHashes.includes(txHash)) {
+          onTxSubmit?.(txHash, pendingTransactions[txHash])
+        }
+      })
+      oldTxs.current = pendingTransactions
+    }
+  }, [pendingTransactions, onTxSubmit])
```

Comment:
using `pendingTransactions` as opposed to txs from `const txs = useAtomValue(transactionsAtom)` to mirror behavior in `onReceipt`

Reply:
nit: can you update the variable names `new...` and `old...` for clarity? `currentPendingTransactions`, `oldPendingTransactions`, etc.